src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1857,126 +1748,68 @@ module Crystal
       end
 
       first_byte = @reader.string.byte_at(start)
-      if first_byte === '+'
-        string_value = "+#{string_value}"
-      elsif first_byte === '-' && num == 0
-        string_value = "-0"
-      end
+      string_value = "+#{string_value}" if first_byte === '+'
 
       @token.type = :NUMBER
       @token.value = string_value
       set_token_raw_from_start(start)
     end
 
     def consume_int_suffix
-      case next_char
-      when '8'
-        next_char
-        @token.number_kind = :i8
-        2
-      when '1'
-        case next_char
-        when '2'
-          if next_char == '8'
-            next_char
-            @token.number_kind = :i128
-            4
-          else
-            raise "invalid int suffix"
-          end
-        when '6'
-          next_char
-          @token.number_kind = :i16
-          3
-        else
-          raise "invalid int suffix"
-        end
-      when '3'
-        if next_char == '2'
-          next_char
-          @token.number_kind = :i32
-          3
-        else
-          raise "invalid int suffix"
-        end
-      when '6'
-        if next_char == '4'
-          next_char
-          @token.number_kind = :i64
-          3
-        else
-          raise "invalid int suffix"
-        end
-      else
-        raise "invalid int suffix"
-      end
+      suffix_info = case next_char
+                    when '1'
+                      case next_char
+                      when '2'
+                        {:i128, 4} if next_char == '8'
+                      when '6'
+                        {:i16, 3}
+                      end
+                    when '3'
+                      {:i32, 3} if next_char == '2'
+                    when '6'
+                      {:i64, 3} if next_char == '4'
+                    when '8'
+                      {:i8, 2}
+                    end
```

Comment:
I believe the previous branch order was on purpose optimized for the (probably) most common cases `u8`/`i8` coming first. Unless there's a strong reason to change this, I would stay with the original order.
Ditto below.

Reply:
On that note, then, the code should be commented to say that the order was chosen for the specified reason.